### PR TITLE
Sync brightness

### PR DESCRIPTION
**Pull request by ufo22940268**
Sync brightness of internal monitor to  external monitor instantly.

This feature will bring following benefits

1. You only need to manage one monitor brightness.
2. If you enable the ambient light sensor of macbook or imac, the external monitor will change according the light at the same time.

### CODE REVIEW

**Comment by ricardopolo** [comment]
:shipit:

**Comment by mafredri** [comment]
This would be a very nice addition to this project, good job @ufo22940268.

One feature that would be nice to see as part of this would be the ability to define the brightness scale non-linearly. I'd think it's very rare that the brightness of two different monitors scale linearly and min/max is not something that can account for that.

Take my DELL U2715H for instance, here's a table of for its brightness values that match my MacBook internal display quite well (using ocular inspection).

| DELL U2715H | MacBook Internal Display |
| :---------: | :----------------------: |
|    0 - 9    |         35 - 47          |
|   10 - 19   |         60 - 69          |
|   20 - 49   |        69 - 88.33        |
|  50 - 100   |        85 - 97.75        |

**Comment by pvaibhav** [comment]
> One feature that would be nice to see as part of this would be the ability to define the brightness scale non-linearly. 

Check how [Lunar](https://lunar.fyi) does it by defining some scaling constants:

CleanShot 2020-10-28 at 11 40 25@2x
CleanShot 2020-10-28 at 11 41 05@2x

**Comment by ricardopolo** [comment]
But please don't be like Lunar. Lunar lets you change the brightness of each display independently, making not very usefull the sync option. A great thing about Monitor Control is that when you change the brightness with your keyboard, it changes it in all your monitor at the same time.

File: MonitorControl.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -447,24 +457,6 @@
...
-			);
-			name = "[Format] Run SwiftFormat";
-			outputFileListPaths = (
-			);
-			outputPaths = (
-			);
-			runOnlyForDeploymentPostprocessing = 0;
-			shellPath = /bin/sh;
-			shellScript = "if which swiftformat >/dev/null; then\n  swiftformat .\nelse\n  echo \"warning: SwiftFormat not installed, download from https://github.com/nicklockwood/SwiftFormat\" >&2\nfi\n";
-		};
```
Comment (JoniVR) [inline MonitorControl.xcodeproj/project.pbxproj:467]: Please revert this change.

File: MonitorControl.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -501,20 +493,6 @@
...
-			);
-			inputPaths = (
-			);
-			name = "[Lint] Run SwiftLint";
-			outputPaths = (
-			);
-			runOnlyForDeploymentPostprocessing = 0;
-			shellPath = /bin/sh;
-			shellScript = "if which swiftlint >/dev/null; then\n  swiftlint\nelse\n  echo \"warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint\" >&2\nfi\n";
-		};
```
Comment (JoniVR) [inline MonitorControl.xcodeproj/project.pbxproj:517]: Please revert this change.

File: MonitorControl.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -337,8 +349,6 @@
 			isa = PBXNativeTarget;
 			buildConfigurationList = 56754EB81D9A4016007BCDC5 /* Build configuration list for PBXNativeTarget "MonitorControl" */;
 			buildPhases = (
-				28D1DE0C227FCFAF004CB494 /* [Format] Run SwiftFormat */,
-				F03A8DF01FFB9D4C0034DC27 /* [Lint] Run SwiftLint */,
```
Comment (JoniVR) [inline MonitorControl.xcodeproj/project.pbxproj:341]: Please revert these.

File: Cartfile [hunk 4]
```diff
@@ -1,4 +1,3 @@
 github "the0neyouseek/MediaKeyTap" "master"
-github "reitermarkus/DDC.swift" "master"
```
Comment (JoniVR) [inline Cartfile:2]: Is there a specific reason why you removed the Carthage dependency here and opted to just include the source? 

It's also throwing an error for me in the import in AppDelegate (No such module 'DDC'). 

Regardless I don't think we should just replace the carthage dependency with a source file as it decreases maintainability.

File: MonitorControl.xcodeproj/xcshareddata/xcschemes/MonitorControl.xcscheme [hunk 5]
```diff
@@ -59,6 +59,13 @@
             ReferencedContainer = "container:MonitorControl.xcodeproj">
          </BuildableReference>
       </BuildableProductRunnable>
+      <EnvironmentVariables>
+         <EnvironmentVariable
+            key = "skip_read_priviledges"
+            value = "true"
+            isEnabled = "YES">
+         </EnvironmentVariable>
+      </EnvironmentVariables>
```
Comment (JoniVR) [inline MonitorControl.xcodeproj/xcshareddata/xcschemes/MonitorControl.xcscheme:68]: What does this do exactly (I'm not familiar)? Also I think there's a typo in `skip_read_priviledges`? Shouldn't it be `skip_read_privileges`?

**Review by JoniVR — CHANGES_REQUESTED** [review]
Hi,

Thanks for the PR, I added some comments specific to the code which need changes. 

Please also fix formatting and indentation generally. As you seem to have removed the formatter for this.

**Comment by JoniVR** [comment]
@ricardopolo I'm not too familiar with Lunar except I've seen it mentioned a few times here. Could you please elaborate? Disabling the sync behaviour would just use the current behaviour I'd think?

File: Cartfile (hunk 4 above)
Comment (JoniVR) [reply Cartfile:2]: I see the issue now, it's not working properly without doing it this way.. Still, we should probably look into getting it fixed instead of just manually including it.

**Comment by ricardopolo** [comment]
@JoniVR Lunar changes the brightness of each screen separately. So the whole "sync" thing is not very useful. If you use your keyboard to change the brightness it mess all.

**Comment by alex955** [comment]
This feature would be so awesome. @ufo22940268 Are you still working on the requested changes?

**Comment by ufo22940268** [comment]
I am too busy on working. Don't have time to complete this PR. Sorry

**Comment by pvaibhav** [comment]
@JoniVR  sorry I never replied to your questions. I'm not too familiar with AppKit and constraints / UI coding etc. so most of your questions I have no idea how to respond.

Anyway, I've since moved on to my own command line script which reads the ambient lighting from an X-rite sensor and uses DDC to set the RGB / contrast / brightness of the monitor to match it. This way I don't have to manually adjust anything.

**Comment by kaeferfreund** [comment]
As far as I can tell, this is not implemented yet?

I have already written a whole app which does exactly the same as monitor control, but since the M1 launch it got useless. This sync feature is the only feature that is missing for me.

Any interest in a new PR based on the latest MC Master Branch?

**Comment by JoniVR** [comment]
@kaeferfreund I think Lunar supports sync mode, so I'd recommend that instead if you want a quick solution. I'm not sure if/when we'll adopt this yet.
